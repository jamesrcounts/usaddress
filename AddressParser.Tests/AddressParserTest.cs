using System.IO;
using System;
using System.Linq;
using System.Text.RegularExpressions;

using ApprovalTests;
using ApprovalTests.Combinations;

using ApprovalUtilities.Utilities;

using Xunit;

namespace USAddress.Tests
{
    public class AddressParserTest
    {
        private static readonly AddressParser Parser = AddressParser.Default;

        [Fact]
        public void StreetLineTests()
        {
            var addressParser = new AddressParser();
            var streetPattern = new Regex(addressParser.StreetPattern, addressParser.MatchOptions);
            var streetPatternMatch = streetPattern.Match("COUNTY ROAD F");
            Assert.NotNull(streetPatternMatch);
            Assert.True(streetPatternMatch.Success);
            Assert.True(streetPatternMatch.Captures.Count == 1);
            Assert.True(streetPatternMatch.Captures[0].Value == "COUNTY ROAD F");
        }

        [Fact]
        public void ParseAddressLine()
        {
            var addressParser = new AddressParser();
            var result = addressParser.ParseAddressLine("3360 County Road F", true);
            Approvals.Verify(result);
        }

        [Fact]
        public void StateDoesNotMatchCity()
        {
            var reg = new Regex(Parser.PlacePattern, Parser.MatchOptions);
            var match = reg.Match("URB LA FABRICA,PR,00704");
            var state = match.Groups[Components.State].Value;
            Approvals.Verify(state);
        }

        [Fact]
        public void AddressRegexIsLazy()
        {
            var addressParser = new AddressParser();
            addressParser.StreetSuffixes.Add("EDGE", "EDGE");
            Approvals.Verify(addressParser.AddressRegex);
        }

        /// <summary>
        /// People in Wisconsin and Illinois are eating too much cheese, apparently, because
        /// you can encounter house numbers with letters in them. It's similar to the
        /// Utah grid-system, except the gridness is all crammed into the house number.
        /// </summary>
        [Fact]
        public void CanParseAddressWithAlphanumericRange()
        {
            Approvals.Verify(Parser.ParseAddress("N6W23001 BLUEMOUND ROAD, ROLLING MEADOWS, IL, 12345"));
        }

        /// <summary>
        /// In Virginia Beach, for example, there's a South Blvd, which could really
        /// throw a spanner into our predirectional/postdirectional parsing. We call
        /// this case out specifically in our regex.
        /// </summary>
        [Fact]
        public void CanParseAddressWithCardinalStreetName()
        {
            Approvals.Verify(Parser.ParseAddress("500 SOUTH STREET VIRGINIA BEACH VIRGINIA 23452"));
        }

        [Fact]
        public void CanParseAddressWithoutPunctuation()
        {
            Approvals.Verify(Parser.ParseAddress("999 West 89th Street Apt A New York NY 10024"));
        }

        /// <summary>
        /// When people live in apartments with letters, they sometimes attach the apartment
        /// letter to the end of the house number. This is wrong, and these people need to be
        /// lined up and individually slapped. We pull out the unit and designate it as "APT",
        /// which in my experience is the designator that USPS uses in the vast, vast majority
        /// of cases.
        /// </summary>
        [Fact]
        public void CanParseAddressWithRangedUnitAttachedToNumber()
        {
            Approvals.Verify(Parser.ParseAddress("403D BERRYFIELD LANE CHESAPEAKE VA 23224"));
        }

        [Fact]
        public void CanParseAddressWithRangelessSecondaryUnit()
        {
            Approvals.Verify(Parser.ParseAddress("1050 Broadway Penthouse, New York, NY 10001"));
        }

        /// <summary>
        /// Speaking of weird addresses, sometimes people put a space in the number.
        /// USPS says we should squash it together.
        /// </summary>
        [Fact]
        public void CanParseAddressWithSpacedAlphanumericRange()
        {
            Approvals.Verify(Parser.ParseAddress("N645 W23001 BLUEMOUND ROAD, ROLLING MEADOWS, IL, 12345"));
        }

        /// <summary>
        /// At least it's not platform 9 3/4.
        /// </summary>
        [Fact]
        public void CanParseFractionalAddress()
        {
            Approvals.Verify(Parser.ParseAddress("123 1/2 MAIN ST, RICHMOND, VA 23221"));
        }

        /// <summary>
        /// Grid-style addresses are common in parts of Utah. The official USPS address database
        /// in this case treats "E" as a pre-directional, "1700" as the street name, and "S" as a
        /// post-directional, and nothing as the suffix, so that's how we parse it, too.
        /// </summary>
        [Fact]
        public void CanParseGridStyleAddress()
        {
            Approvals.Verify(Parser.ParseAddress("842 E 1700 S, Salt Lake City, UT 84105"));
        }

        /// <summary>
        /// Military addresses seem to follow no convention whatsoever in the
        /// street line, but the APO/FPO/DPO AA/AE/AP 9NNNN part of the place line
        /// is pretty well standardized. I've made a special exception for these
        /// kinds of addresses so that the street line is just dumped as-is into
        /// the StreetLine field.
        /// </summary>
        [Fact]
        public void CanParseMilitaryAddress()
        {
            Approvals.Verify(Parser.ParseAddress("PSC BOX 453, APO AE 99969"));
        }

        [Fact]
        public void CanParsePostOfficeBoxAddress()
        {
            Approvals.Verify(Parser.ParseAddress("P.O. BOX 4857, New York, NY 10001"));
        }

        /// <summary>
        /// In parts of New York City, some people feel REALLY STRONGLY about
        /// the hyphen in their house number. The numbering system makes sense,
        /// but the USPS address database doesn't support hyphens in the number field.
        /// To the USPS, the hyphen does not exist, but the DMM specifically does say
        /// that "if present, the hyphen should not be removed."
        /// </summary>
        [Fact]
        public void CanParseQueensStyleAddress()
        {
            Approvals.Verify(Parser.ParseAddress("123-465 34th St New York NY 12345"));
        }

        [Fact]
        public void CanParseTypicalAddressWithoutPunctuationAfterStreetLine()
        {
            Approvals.Verify(Parser.ParseAddress("1005 N Gravenstein Highway Sebastopol, CA 95472"));
        }

        [Fact]
        public void CanParseTypicalAddressWithPunctuation()
        {
            Approvals.Verify(Parser.ParseAddress("1005 N Gravenstein Highway, Sebastopol, CA 95472"));
        }

        [Fact]
        public void MatchSecondaryUnitAsWord()
        {
            var streets = new[] { "9999 RIVERSIDE DR WEST #8B", "9999 RIVERSIDE DR WEST SIDE LEFT", };
            var results = streets.Select(s => new { Text = s, Match = Regex.Match(s, AddressParser.Default.RangelessSecondaryUnitPattern, AddressParser.Default.MatchOptions) });
            Approvals.VerifyAll(results, "Match");
        }

        [Fact]
        public void ParseExampleAddresses()
        {
            var text = File.ReadAllText(PathUtilities.GetAdjacentFile("samples.txt"));
            var examples = text.Split(new[] {"\r\n", "\n"}, StringSplitOptions.RemoveEmptyEntries)
                .Where(s => !s.StartsWith("#", StringComparison.OrdinalIgnoreCase));
            bool[] normalize = { true, false };

            CombinationApprovals.VerifyAllCombinations(
                Parser.ParseAddress,
                x => Environment.NewLine + ((AddressParseResult)x).WritePropertiesToString(),
                examples,
                normalize);
        }

        [Fact]
        public void ParseWithoutNormalization()
        {
            Approvals.Verify(Parser.ParseAddress("999 West 89th Street Apt A New York NY 10024", false));
        }

        [Fact]
        public void VerifyAllSecondaryUnitPattern()
        {
            Approvals.Verify(AddressParser.Default.AllSecondaryUnitPattern);
        }

        [Fact]
        public void VerifyAllUnits()
        {
            Approvals.VerifyAll(AddressParser.Default.AllUnits);
        }

        [Fact]
        public void VerifyCityAndStatePattern()
        {
            Approvals.Verify(AddressParser.Default.CityAndStatePattern);
        }

        [Fact]
        public void VerifyDirectionalNames()
        {
            Approvals.VerifyAll(AddressParser.Default.DirectionalNames);
        }

        [Fact]
        public void VerifyDirectionalPattern()
        {
            Approvals.Verify(AddressParser.Default.DirectionalPattern);
        }

        [Fact]
        public void VerifyMatchOptions()
        {
            Approvals.Verify(AddressParser.Default.MatchOptions);
        }

        [Fact]
        public void VerifyPlacePattern()
        {
            Approvals.Verify(AddressParser.Default.PlacePattern);
        }

        [Fact]
        public void VerifyPostalBoxPattern()
        {
            Approvals.Verify(AddressParser.Default.PostalBoxPattern);
        }

        [Fact]
        public void VerifyRangedSecondaryUnitPattern()
        {
            Approvals.Verify(AddressParser.Default.RangedSecondaryUnitPattern);
        }

        [Fact]
        public void VerifyRangedUnits()
        {
            Approvals.VerifyAll(AddressParser.Default.RangedUnits);
        }

        [Fact]
        public void VerifyRangelessSecondaryUnitPattern()
        {
            Approvals.Verify(AddressParser.Default.RangelessSecondaryUnitPattern);
        }

        [Fact]
        public void VerifyRangelessUnits()
        {
            Approvals.VerifyAll(AddressParser.Default.RangelessUnits);
        }

        [Fact]
        public void VerifyStatePattern()
        {
            Approvals.Verify(AddressParser.Default.StatePattern);
        }

        [Fact]
        public void VerifyStates()
        {
            Approvals.VerifyAll(AddressParser.Default.StatesAndProvinces);
        }

        [Fact]
        public void VerifyStreetPattern()
        {
            Approvals.Verify(AddressParser.Default.StreetPattern);
        }

        [Fact]
        public void VerifySuffixes()
        {
            Approvals.VerifyAll(AddressParser.Default.StreetSuffixes);
        }

        [Fact]
        public void VerifySuffixPattern()
        {
            Approvals.Verify(AddressParser.Default.SuffixPattern);
        }

        [Fact]
        public void VerifyZipPattern()
        {
            Approvals.Verify(AddressParser.ZipPattern);
        }
    }
}