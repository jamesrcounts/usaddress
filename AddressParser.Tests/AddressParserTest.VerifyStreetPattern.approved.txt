
                        (?:
                          # special case for addresses like 100 COUNTY ROAD F  http://regexstorm.net/tester?p=%28%3f%3a%28%3f%3cSTREET%3eCOUNTY%5cW%2bROAD%5cW*%5cw%2b%29%29&i=COUNTY+ROAD+FG%0d%0a%0d%0a&o=ixs
                          (?:(?<STREET>COUNTY\W*ROAD\W*\w+))
                          |
                          # special case for addresses like 100 South Street
                          (?:(?<STREET>NORTHEAST|SOUTHEAST|SOUTHWEST|NORTHWEST|N\.E\.|S\.E\.|S\.W\.|N\.W\.|NORTH|SOUTH|EAST|WEST|N\.|E\.|S\.|W\.|NE|SE|SW|NW|N|E|S|W)\W+
                             (?<SUFFIX>EXPRESSWAY|EXTENSIONS|THROUGHWAY|TRAFFICWAY|BOULEVARD|CROSSROAD|EXTENSION|JUNCTIONS|MOUNTAINS|STRAVENUE|UNDERPASS|CAUSEWAY|CRESCENT|CROSSING|JUNCTION|MOTORWAY|MOUNTAIN|OVERPASS|PARKWAYS|TURNPIKE|VILLAGES|VILLIAGE|CAUSWAY|CENTERS|CIRCLES|CORNERS|CRECENT|CRESENT|CRSSING|ESTATES|EXPRESS|FORESTS|FREEWAY|GARDENS|GATEWAY|HARBORS|HEIGHTS|HIGHWAY|HOLLOWS|ISLANDS|JUNCTON|LANDING|MEADOWS|MISSION|MOUNTIN|ORCHARD|PARKWAY|PASSAGE|PLAINES|PRAIRIE|RANCHES|SPRINGS|SQUARES|STATION|STRAVEN|STREETS|STRVNUE|TERRACE|TUNNELS|VALLEYS|VIADUCT|VILLAGE|ARCADE|AVENUE|BLUFFS|BOTTOM|BRANCH|BRIDGE|BROOKS|BYPASS|CANYON|CENTER|CENTRE|CIRCLE|CLIFFS|COMMON|CORNER|COURSE|COURTS|CRSCNT|CRSENT|CRSSNG|DIVIDE|DRIVES|ESTATE|EXTNSN|FIELDS|FOREST|FORGES|FREEWY|GARDEN|GATEWY|GATWAY|GREENS|GROVES|HARBOR|HEIGHT|HIGHWY|HOLLOW|ISLAND|ISLNDS|JCTION|JUNCTN|KNOLLS|LIGHTS|MANORS|MEADOW|MEDOWS|MNTAIN|ORCHRD|PARKWY|PLAINS|POINTS|PRARIE|RADIAL|RADIEL|RAPIDS|RIDGES|SHOALS|SHOARS|SHORES|SKYWAY|SPRING|SPRNGS|SQUARE|STRAVE|STRAVN|STREAM|STREET|STREME|SUMITT|SUMMIT|TRACES|TRACKS|TRAILS|TUNNEL|TURNPK|UNIONS|VALLEY|VIADCT|VILLAG|ALLEE|ALLEY|ANNEX|AVENU|AVNUE|BAYOO|BAYOU|BEACH|BLUFF|BOTTM|BOULV|BRDGE|BRNCH|BROOK|BURGS|BYPAS|CANYN|CENTR|CIRCL|CLIFF|CNTER|COURT|COVES|CRCLE|CREEK|CREST|CRSNT|CURVE|DRIVE|FALLS|FERRY|FIELD|FLATS|FORDS|FORGE|FORKS|FRWAY|GARDN|GLENS|GRDEN|GRDNS|GREEN|GROVE|GTWAY|HARBR|HAVEN|HILLS|HIWAY|HOLWS|HRBOR|INLET|ISLES|ISLND|JCTNS|KNOLL|LAKES|LANES|LIGHT|LNDNG|LOCKS|LODGE|LOOPS|MANOR|MILLS|MISSN|MNTNS|MOUNT|PARKS|PATHS|PIKES|PINES|PKWAY|PKWYS|PLACE|PLAIN|PLAZA|POINT|PORTS|RANCH|RAPID|RIDGE|RIVER|RNCHS|ROADS|ROUTE|SHOAL|SHOAR|SHORE|SPNGS|SPRNG|SPURS|STATN|STRAV|STRVN|SUMIT|TRACE|TRACK|TRAIL|TRNPK|TUNEL|TUNLS|TUNNL|UNION|VALLY|VIEWS|VILLE|VILLG|VISTA|WALKS|WELLS|BLFS|BLVD|BRKS|CSWY|CTRS|CIRS|CLFS|CORS|CRSE|CRES|CRST|XING|CURV|ESTS|EXPY|EXTS|FLDS|FLTS|FRDS|FRST|FRGS|FRKS|GDNS|GTWY|GLNS|GRNS|GRVS|HBRS|HOLW|INLT|ISLE|JCTS|KNLS|LNDG|LGTS|LCKS|LOOP|MNRS|MDWS|MTNS|MTWY|ORCH|OPAS|OVAL|PARK|PKWY|PSGE|PATH|PIKE|PNES|PLNS|PRTS|RADL|RNCH|RPDS|RDGS|SHLS|SHRS|SKWY|SPGS|SPUR|STRA|STRM|TRWY|TPKE|TRCE|TRAK|TRFY|TUNL|UPAS|VLYS|VLGS|WALK|ALLY|ANEX|ANNX|AVEN|BEND|BLUF|BOUL|BURG|BYPA|BYPS|CAMP|CAPE|CENT|CIRC|CLUB|CNTR|CNYN|COVE|CRCL|DALE|DRIV|EXPR|EXPW|EXTN|FLAT|FORD|FORG|FORK|FORT|FRRY|FRWY|GLEN|GRDN|GROV|HARB|HAVN|HGTS|HILL|HIWY|HLLW|HWAY|JCTN|KEYS|KNOL|LAKE|LANE|LDGE|LOAF|LOCK|LODG|MILL|MNTN|MSSN|MTIN|NECK|PINE|PLZA|PORT|RDGE|REST|RIVR|ROAD|SPNG|SQRE|SQRS|STRT|TERR|TRKS|TRLS|TRPK|VDCT|VIEW|VILL|VIST|VLLY|VSTA|WELL|ALY|ANX|ARC|AVE|BYU|BCH|BND|BLF|BTM|BRG|BRK|BGS|BYP|CYN|CPE|CTR|CIR|CRK|CLF|CLB|CMN|COR|CTS|CVS|XRD|DRS|EST|EXT|FLS|FRY|FLD|FLT|FRD|FRG|FRK|FWY|GDN|GLN|GRN|GRV|HBR|HVN|HTS|HWY|HLS|ISS|JCT|KYS|KNL|LKS|LDG|LGT|LCK|MNR|MDW|MLS|MSN|MTN|NCK|PNE|PLN|PLZ|PTS|PRT|RPD|RDG|RST|RIV|RDS|RTE|SHL|SHR|SPG|SQS|STA|STS|SMT|TER|TRL|UNS|VLY|VIA|VWS|VLG|VIS|WLS|WAY|AVN|BOT|CEN|CMP|CRT|DAM|DIV|DRV|DVD|EXP|FRT|KEY|MNT|OVL|PKY|PRK|PRR|RAD|RVR|SQR|SQU|STN|STR|TPK|TRK|VST|BR|BG|CP|CT|CV|DL|DM|DV|DR|FT|HL|IS|KY|LN|LK|LF|ML|MT|PL|PT|PR|RD|SQ|ST|UN|VW|VL|WL|AV|CK|CR|HT|LA|PK|TR|WY)\b)
                          |
                          (?:(?<PREDIRECTIONAL>NORTHEAST|SOUTHEAST|SOUTHWEST|NORTHWEST|N\.E\.|S\.E\.|S\.W\.|N\.W\.|NORTH|SOUTH|EAST|WEST|N\.|E\.|S\.|W\.|NE|SE|SW|NW|N|E|S|W)\W+)?
                          (?:
                            (?<STREET>[^,]*\d)
                            (?:[^\w,]*(?<POSTDIRECTIONAL>NORTHEAST|SOUTHEAST|SOUTHWEST|NORTHWEST|N\.E\.|S\.E\.|S\.W\.|N\.W\.|NORTH|SOUTH|EAST|WEST|N\.|E\.|S\.|W\.|NE|SE|SW|NW|N|E|S|W)\b)
                           |
                            (?<STREET>[^,]+)
                            (?:[^\w,]+(?<SUFFIX>EXPRESSWAY|EXTENSIONS|THROUGHWAY|TRAFFICWAY|BOULEVARD|CROSSROAD|EXTENSION|JUNCTIONS|MOUNTAINS|STRAVENUE|UNDERPASS|CAUSEWAY|CRESCENT|CROSSING|JUNCTION|MOTORWAY|MOUNTAIN|OVERPASS|PARKWAYS|TURNPIKE|VILLAGES|VILLIAGE|CAUSWAY|CENTERS|CIRCLES|CORNERS|CRECENT|CRESENT|CRSSING|ESTATES|EXPRESS|FORESTS|FREEWAY|GARDENS|GATEWAY|HARBORS|HEIGHTS|HIGHWAY|HOLLOWS|ISLANDS|JUNCTON|LANDING|MEADOWS|MISSION|MOUNTIN|ORCHARD|PARKWAY|PASSAGE|PLAINES|PRAIRIE|RANCHES|SPRINGS|SQUARES|STATION|STRAVEN|STREETS|STRVNUE|TERRACE|TUNNELS|VALLEYS|VIADUCT|VILLAGE|ARCADE|AVENUE|BLUFFS|BOTTOM|BRANCH|BRIDGE|BROOKS|BYPASS|CANYON|CENTER|CENTRE|CIRCLE|CLIFFS|COMMON|CORNER|COURSE|COURTS|CRSCNT|CRSENT|CRSSNG|DIVIDE|DRIVES|ESTATE|EXTNSN|FIELDS|FOREST|FORGES|FREEWY|GARDEN|GATEWY|GATWAY|GREENS|GROVES|HARBOR|HEIGHT|HIGHWY|HOLLOW|ISLAND|ISLNDS|JCTION|JUNCTN|KNOLLS|LIGHTS|MANORS|MEADOW|MEDOWS|MNTAIN|ORCHRD|PARKWY|PLAINS|POINTS|PRARIE|RADIAL|RADIEL|RAPIDS|RIDGES|SHOALS|SHOARS|SHORES|SKYWAY|SPRING|SPRNGS|SQUARE|STRAVE|STRAVN|STREAM|STREET|STREME|SUMITT|SUMMIT|TRACES|TRACKS|TRAILS|TUNNEL|TURNPK|UNIONS|VALLEY|VIADCT|VILLAG|ALLEE|ALLEY|ANNEX|AVENU|AVNUE|BAYOO|BAYOU|BEACH|BLUFF|BOTTM|BOULV|BRDGE|BRNCH|BROOK|BURGS|BYPAS|CANYN|CENTR|CIRCL|CLIFF|CNTER|COURT|COVES|CRCLE|CREEK|CREST|CRSNT|CURVE|DRIVE|FALLS|FERRY|FIELD|FLATS|FORDS|FORGE|FORKS|FRWAY|GARDN|GLENS|GRDEN|GRDNS|GREEN|GROVE|GTWAY|HARBR|HAVEN|HILLS|HIWAY|HOLWS|HRBOR|INLET|ISLES|ISLND|JCTNS|KNOLL|LAKES|LANES|LIGHT|LNDNG|LOCKS|LODGE|LOOPS|MANOR|MILLS|MISSN|MNTNS|MOUNT|PARKS|PATHS|PIKES|PINES|PKWAY|PKWYS|PLACE|PLAIN|PLAZA|POINT|PORTS|RANCH|RAPID|RIDGE|RIVER|RNCHS|ROADS|ROUTE|SHOAL|SHOAR|SHORE|SPNGS|SPRNG|SPURS|STATN|STRAV|STRVN|SUMIT|TRACE|TRACK|TRAIL|TRNPK|TUNEL|TUNLS|TUNNL|UNION|VALLY|VIEWS|VILLE|VILLG|VISTA|WALKS|WELLS|BLFS|BLVD|BRKS|CSWY|CTRS|CIRS|CLFS|CORS|CRSE|CRES|CRST|XING|CURV|ESTS|EXPY|EXTS|FLDS|FLTS|FRDS|FRST|FRGS|FRKS|GDNS|GTWY|GLNS|GRNS|GRVS|HBRS|HOLW|INLT|ISLE|JCTS|KNLS|LNDG|LGTS|LCKS|LOOP|MNRS|MDWS|MTNS|MTWY|ORCH|OPAS|OVAL|PARK|PKWY|PSGE|PATH|PIKE|PNES|PLNS|PRTS|RADL|RNCH|RPDS|RDGS|SHLS|SHRS|SKWY|SPGS|SPUR|STRA|STRM|TRWY|TPKE|TRCE|TRAK|TRFY|TUNL|UPAS|VLYS|VLGS|WALK|ALLY|ANEX|ANNX|AVEN|BEND|BLUF|BOUL|BURG|BYPA|BYPS|CAMP|CAPE|CENT|CIRC|CLUB|CNTR|CNYN|COVE|CRCL|DALE|DRIV|EXPR|EXPW|EXTN|FLAT|FORD|FORG|FORK|FORT|FRRY|FRWY|GLEN|GRDN|GROV|HARB|HAVN|HGTS|HILL|HIWY|HLLW|HWAY|JCTN|KEYS|KNOL|LAKE|LANE|LDGE|LOAF|LOCK|LODG|MILL|MNTN|MSSN|MTIN|NECK|PINE|PLZA|PORT|RDGE|REST|RIVR|ROAD|SPNG|SQRE|SQRS|STRT|TERR|TRKS|TRLS|TRPK|VDCT|VIEW|VILL|VIST|VLLY|VSTA|WELL|ALY|ANX|ARC|AVE|BYU|BCH|BND|BLF|BTM|BRG|BRK|BGS|BYP|CYN|CPE|CTR|CIR|CRK|CLF|CLB|CMN|COR|CTS|CVS|XRD|DRS|EST|EXT|FLS|FRY|FLD|FLT|FRD|FRG|FRK|FWY|GDN|GLN|GRN|GRV|HBR|HVN|HTS|HWY|HLS|ISS|JCT|KYS|KNL|LKS|LDG|LGT|LCK|MNR|MDW|MLS|MSN|MTN|NCK|PNE|PLN|PLZ|PTS|PRT|RPD|RDG|RST|RIV|RDS|RTE|SHL|SHR|SPG|SQS|STA|STS|SMT|TER|TRL|UNS|VLY|VIA|VWS|VLG|VIS|WLS|WAY|AVN|BOT|CEN|CMP|CRT|DAM|DIV|DRV|DVD|EXP|FRT|KEY|MNT|OVL|PKY|PRK|PRR|RAD|RVR|SQR|SQU|STN|STR|TPK|TRK|VST|BR|BG|CP|CT|CV|DL|DM|DV|DR|FT|HL|IS|KY|LN|LK|LF|ML|MT|PL|PT|PR|RD|SQ|ST|UN|VW|VL|WL|AV|CK|CR|HT|LA|PK|TR|WY)\b)
                            (?:[^\w,]+(?<POSTDIRECTIONAL>NORTHEAST|SOUTHEAST|SOUTHWEST|NORTHWEST|N\.E\.|S\.E\.|S\.W\.|N\.W\.|NORTH|SOUTH|EAST|WEST|N\.|E\.|S\.|W\.|NE|SE|SW|NW|N|E|S|W)\b)?
                           |
                            (?<STREET>[^,]+?)
                            (?:[^\w,]+(?<SUFFIX>EXPRESSWAY|EXTENSIONS|THROUGHWAY|TRAFFICWAY|BOULEVARD|CROSSROAD|EXTENSION|JUNCTIONS|MOUNTAINS|STRAVENUE|UNDERPASS|CAUSEWAY|CRESCENT|CROSSING|JUNCTION|MOTORWAY|MOUNTAIN|OVERPASS|PARKWAYS|TURNPIKE|VILLAGES|VILLIAGE|CAUSWAY|CENTERS|CIRCLES|CORNERS|CRECENT|CRESENT|CRSSING|ESTATES|EXPRESS|FORESTS|FREEWAY|GARDENS|GATEWAY|HARBORS|HEIGHTS|HIGHWAY|HOLLOWS|ISLANDS|JUNCTON|LANDING|MEADOWS|MISSION|MOUNTIN|ORCHARD|PARKWAY|PASSAGE|PLAINES|PRAIRIE|RANCHES|SPRINGS|SQUARES|STATION|STRAVEN|STREETS|STRVNUE|TERRACE|TUNNELS|VALLEYS|VIADUCT|VILLAGE|ARCADE|AVENUE|BLUFFS|BOTTOM|BRANCH|BRIDGE|BROOKS|BYPASS|CANYON|CENTER|CENTRE|CIRCLE|CLIFFS|COMMON|CORNER|COURSE|COURTS|CRSCNT|CRSENT|CRSSNG|DIVIDE|DRIVES|ESTATE|EXTNSN|FIELDS|FOREST|FORGES|FREEWY|GARDEN|GATEWY|GATWAY|GREENS|GROVES|HARBOR|HEIGHT|HIGHWY|HOLLOW|ISLAND|ISLNDS|JCTION|JUNCTN|KNOLLS|LIGHTS|MANORS|MEADOW|MEDOWS|MNTAIN|ORCHRD|PARKWY|PLAINS|POINTS|PRARIE|RADIAL|RADIEL|RAPIDS|RIDGES|SHOALS|SHOARS|SHORES|SKYWAY|SPRING|SPRNGS|SQUARE|STRAVE|STRAVN|STREAM|STREET|STREME|SUMITT|SUMMIT|TRACES|TRACKS|TRAILS|TUNNEL|TURNPK|UNIONS|VALLEY|VIADCT|VILLAG|ALLEE|ALLEY|ANNEX|AVENU|AVNUE|BAYOO|BAYOU|BEACH|BLUFF|BOTTM|BOULV|BRDGE|BRNCH|BROOK|BURGS|BYPAS|CANYN|CENTR|CIRCL|CLIFF|CNTER|COURT|COVES|CRCLE|CREEK|CREST|CRSNT|CURVE|DRIVE|FALLS|FERRY|FIELD|FLATS|FORDS|FORGE|FORKS|FRWAY|GARDN|GLENS|GRDEN|GRDNS|GREEN|GROVE|GTWAY|HARBR|HAVEN|HILLS|HIWAY|HOLWS|HRBOR|INLET|ISLES|ISLND|JCTNS|KNOLL|LAKES|LANES|LIGHT|LNDNG|LOCKS|LODGE|LOOPS|MANOR|MILLS|MISSN|MNTNS|MOUNT|PARKS|PATHS|PIKES|PINES|PKWAY|PKWYS|PLACE|PLAIN|PLAZA|POINT|PORTS|RANCH|RAPID|RIDGE|RIVER|RNCHS|ROADS|ROUTE|SHOAL|SHOAR|SHORE|SPNGS|SPRNG|SPURS|STATN|STRAV|STRVN|SUMIT|TRACE|TRACK|TRAIL|TRNPK|TUNEL|TUNLS|TUNNL|UNION|VALLY|VIEWS|VILLE|VILLG|VISTA|WALKS|WELLS|BLFS|BLVD|BRKS|CSWY|CTRS|CIRS|CLFS|CORS|CRSE|CRES|CRST|XING|CURV|ESTS|EXPY|EXTS|FLDS|FLTS|FRDS|FRST|FRGS|FRKS|GDNS|GTWY|GLNS|GRNS|GRVS|HBRS|HOLW|INLT|ISLE|JCTS|KNLS|LNDG|LGTS|LCKS|LOOP|MNRS|MDWS|MTNS|MTWY|ORCH|OPAS|OVAL|PARK|PKWY|PSGE|PATH|PIKE|PNES|PLNS|PRTS|RADL|RNCH|RPDS|RDGS|SHLS|SHRS|SKWY|SPGS|SPUR|STRA|STRM|TRWY|TPKE|TRCE|TRAK|TRFY|TUNL|UPAS|VLYS|VLGS|WALK|ALLY|ANEX|ANNX|AVEN|BEND|BLUF|BOUL|BURG|BYPA|BYPS|CAMP|CAPE|CENT|CIRC|CLUB|CNTR|CNYN|COVE|CRCL|DALE|DRIV|EXPR|EXPW|EXTN|FLAT|FORD|FORG|FORK|FORT|FRRY|FRWY|GLEN|GRDN|GROV|HARB|HAVN|HGTS|HILL|HIWY|HLLW|HWAY|JCTN|KEYS|KNOL|LAKE|LANE|LDGE|LOAF|LOCK|LODG|MILL|MNTN|MSSN|MTIN|NECK|PINE|PLZA|PORT|RDGE|REST|RIVR|ROAD|SPNG|SQRE|SQRS|STRT|TERR|TRKS|TRLS|TRPK|VDCT|VIEW|VILL|VIST|VLLY|VSTA|WELL|ALY|ANX|ARC|AVE|BYU|BCH|BND|BLF|BTM|BRG|BRK|BGS|BYP|CYN|CPE|CTR|CIR|CRK|CLF|CLB|CMN|COR|CTS|CVS|XRD|DRS|EST|EXT|FLS|FRY|FLD|FLT|FRD|FRG|FRK|FWY|GDN|GLN|GRN|GRV|HBR|HVN|HTS|HWY|HLS|ISS|JCT|KYS|KNL|LKS|LDG|LGT|LCK|MNR|MDW|MLS|MSN|MTN|NCK|PNE|PLN|PLZ|PTS|PRT|RPD|RDG|RST|RIV|RDS|RTE|SHL|SHR|SPG|SQS|STA|STS|SMT|TER|TRL|UNS|VLY|VIA|VWS|VLG|VIS|WLS|WAY|AVN|BOT|CEN|CMP|CRT|DAM|DIV|DRV|DVD|EXP|FRT|KEY|MNT|OVL|PKY|PRK|PRR|RAD|RVR|SQR|SQU|STN|STR|TPK|TRK|VST|BR|BG|CP|CT|CV|DL|DM|DV|DR|FT|HL|IS|KY|LN|LK|LF|ML|MT|PL|PT|PR|RD|SQ|ST|UN|VW|VL|WL|AV|CK|CR|HT|LA|PK|TR|WY)\b)?
                            (?:[^\w,]+(?<POSTDIRECTIONAL>NORTHEAST|SOUTHEAST|SOUTHWEST|NORTHWEST|N\.E\.|S\.E\.|S\.W\.|N\.W\.|NORTH|SOUTH|EAST|WEST|N\.|E\.|S\.|W\.|NE|SE|SW|NW|N|E|S|W)\b)?
                          )
                        )
                    